src/compiler/scanner.ts

```diff
@@ -2469,55 +2468,92 @@ export function createScanner(languageVersion: ScriptTarget, skipTrivia: boolean
-                    if (flag === undefined) {
-                        error(Diagnostics.Unknown_regular_expression_flag, p, 1);
+            const endOfRegExpBody = pos;
+            if (tokenFlags & TokenFlags.Unterminated) {
+                // Search for the nearest unbalanced bracket for better recovery. Since the expression is
+                // invalid anyways, we take nested square brackets into consideration for the best guess.
+                pos = startOfRegExpBody;
+                inEscape = false;
+                let characterClassDepth = 0;
+                const bracketStack: CharacterCodes[] = [];
+                while (pos < endOfRegExpBody) {
+                    const ch = charCodeUnchecked(pos);
+                    if (inEscape) {
+                        inEscape = false;
                     }
-                    else if (regExpFlags & flag) {
-                        error(Diagnostics.Duplicate_regular_expression_flag, p, 1);
+                    else if (ch === CharacterCodes.backslash) {
+                        inEscape = true;
                     }
-                    else if (((regExpFlags | flag) & RegularExpressionFlags.UnicodeMode) === RegularExpressionFlags.UnicodeMode) {
-                        error(Diagnostics.The_Unicode_u_flag_and_the_Unicode_Sets_v_flag_cannot_be_set_simultaneously, p, 1);
+                    else if (ch === CharacterCodes.openBracket) {
+                        characterClassDepth++;
                     }
-                    else {
-                        regExpFlags |= flag;
-                        checkRegularExpressionFlagAvailable(flag, p);
-                    }
-                }
-                p++;
-            }
-            pos = p;
-            if (reportErrors) {
-                const saveTokenStart = tokenStart;
-                const saveTokenFlags = tokenFlags;
-                const savePos = pos;
-                const saveEnd = end;
-                pos = tokenStart + 1;
-                end = endOfBody;
-                scanRegularExpressionWorker(regExpFlags, isUnterminated, /*annexB*/ true);
-                tokenStart = saveTokenStart;
-                tokenFlags = saveTokenFlags;
-                pos = savePos;
-                end = saveEnd;
+                    else if (ch === CharacterCodes.closeBracket && characterClassDepth) {
+                        characterClassDepth--;
+                    }
+                    else if (!characterClassDepth) {
+                        if (ch === CharacterCodes.openParen) {
+                            bracketStack.push(CharacterCodes.closeParen);
+                        }
+                        else if (ch === CharacterCodes.openBrace) {
+                            bracketStack.push(CharacterCodes.closeBrace);
+                        }
+                        else if (ch === lastOrUndefined(bracketStack)) {
+                            bracketStack.pop();
+                        }

```

Comment:
Parentheses and braces are not part of the [RegularExpressionBody](https://tc39.es/ecma262/#prod-RegularExpressionBody) lexical grammar, and thus unbalanced parentheses don't affect the extent of a [RegularExpressionLiteral](https://tc39.es/ecma262/#prod-RegularExpressionLiteral). If they are unbalanced, that doesn't mean the regexp is considered to be unterminated.

Reply:
But this section is all about recovering an unterminated regex. This section does not apply to well-terminated regexes, regardless of whether they’re balanced or unbalanced.
And, aren’t parentheses and braces [RegularExpressionNonTerminator](https://tc39.es/ecma262/#prod-RegularExpressionNonTerminator)s, which are in turn parts of RegularExpressionBody?

Reply:
They are part of the body, but RegularExpressionBody itself doesn't care if they are balanced. That check occurs later as part of the _Pattern_ grammar, not the lexical grammar. That said, I can see why they need to be considered for recovery purposes.